### PR TITLE
Add `InArrayValidator`

### PR DESCRIPTION
**Pull request by sukhwinder33445**

### CODE REVIEW

File: src/InArrayValidator.php [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+     * Validates whether the value exists in given haystack
+     *
+     * Required options:
+     *
+     * - haystack: (array) Haystack of possible values
+     *
+     * Optional options:
+     *
+     * - strict: (bool) Whether the types of the needle in the haystack should also match, default false
+     *
```
Comment (nilmerg) [inline src/InArrayValidator.php:33]: Why this empty line?

File: src/InArrayValidator.php [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            }
+
+            if (empty($notInArrayValues)) {
+                return true;
+            }
+
+            $value = implode(', ', $notInArrayValues);
+            if (count($notInArrayValues) > 1) {
+                $this->addMessage(sprintf(
+                    $this->translate("'%s' were not found in the haystack"), //BETTER CODE REQUIRED HERE
```
Comment (nilmerg) [inline src/InArrayValidator.php:143]: This could be an option for `translatePlural`.

File: tests/InArrayValidatorTest.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+{
+    public function testInArrayValidatorWithValidValue()
+    {
+        StaticTranslator::$instance = new NoopTranslator();
+
+        $validator = new InArrayValidator([
+            'haystack' => ['test', 'foo', 'bar', 5, [], 0.008]
+        ]);
+
+        $this->assertTrue($validator->isValid('foo'), 'foo exists in the given haystack');
```
Comment (nilmerg) [inline tests/InArrayValidatorTest.php:19]: I'm quite sure the message is used if the assertion fails. So, `foo not found in ..` is correct.

Comment (nilmerg) [reply tests/InArrayValidatorTest.php:19]: This seems to apply to all of your messages, in all test cases. Please adjust all of them.

**Comment by nilmerg** [comment]
If that's the only reason, I'd like to have that removed. The `recursive` option of said validator is not in use anywhere.

File: src/InArrayValidator.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+        return $this;
+    }
+
+    public function isValid($value)
+    {
+        // Multiple isValid() calls must not stack validation messages
+        $this->clearMessages();
+
+        $haystack = $this->getHaystack();
```
Comment (lippserd) [inline src/InArrayValidator.php:89]: What about something like this?

```
// Class method:
$looseEqual = function ($a, $b): bool {
    return $a == $b;
};

// Class method:
$strictEqual = function ($a, $b): bool {
    return $a === $b;
};

$notInArray = array_udiff((array) $value, $haystack, $strict ? $strictEqual : $looseEqual);
```

Supports strict, non-arrays and arrays as needle and is more readable.

**Review by nilmerg — APPROVED** [review]

File: src/InArrayValidator.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace ipl\Validator;
+
+use ipl\I18n\Translation;
+
+class InArrayValidator extends BaseValidator
```
Comment (lippserd) [inline src/InArrayValidator.php:7]: Add PHPDoc, e.g.

> Validate if specific single or multiple values exist in an array

File: src/InArrayValidator.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    use Translation;
+
+    /** @var array The array */
+    protected $haystack;
+
+    /** @var bool Whether the types of the needle in the haystack should also match */
+    protected $strict = false;
+
+    /**
+     * Validates whether the value exists in given haystack
```
Comment (lippserd) [inline src/InArrayValidator.php:18]: The constructor does not validate. And you can use Markdown. Once we render API docs, it makes sense to have them good looking. PHPStorm also renders Markdown:

```
Create a new validator with the given options

**Supported options:**

* `haystack`: (`array`) The array
* `strict`: (`bool`) Whether ..., default `false`

File: src/InArrayValidator.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    /**
+     * Returns the haystack
+     *
+     * @return array
+     */
+    public function getHaystack(): array
+    {
+        return $this->haystack ?? [];
```
Comment (lippserd) [inline src/InArrayValidator.php:43]: Let `haystack` default to the empty array instead.

File: src/InArrayValidator.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    /**
+     * Set the haystack
+     *
+     * @param  array $haystack
```
Comment (lippserd) [inline src/InArrayValidator.php:49]: Too much whitespaces after `@param`. Should be just one.

File: src/InArrayValidator.php [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    public function isStrict(): bool
+    {
+        return $this->strict;
+    }
+
+    /**
+     * Set whether the types of the needle in the haystack should also match
+     *
+     * @param  bool $strict
```
Comment (lippserd) [inline src/InArrayValidator.php:73]: Too much whitespaces after `@param`. Should be just one.

File: src/InArrayValidator.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        return $this->haystack ?? [];
+    }
+
+    /**
+     * Set the haystack
+     *
+     * @param  array $haystack
+     *
+     * @return self
```
Comment (lippserd) [inline src/InArrayValidator.php:51]: It's `@return $this`.

File: src/InArrayValidator.php [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        return $this->strict;
+    }
+
+    /**
+     * Set whether the types of the needle in the haystack should also match
+     *
+     * @param  bool $strict
+     *
+     * @return self
```
Comment (lippserd) [inline src/InArrayValidator.php:75]: It's `@return $this`.

File: src/InArrayValidator.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        if (isset($options['haystack'])) {
+            $this->setHaystack($options['haystack']);
+        }
+
+        $this->setStrict($options['strict'] ?? false);
+    }
+
+    /**
+     * Returns the haystack
```
Comment (lippserd) [inline src/InArrayValidator.php:37]: Just `Get the haystack`.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: src/InArrayValidator.php (hunk 4 above)
Comment (nilmerg) [reply src/InArrayValidator.php]: There's the spaceship operator to achieve this: `<=>`
There's no strict version though.

Comment (lippserd) [reply src/InArrayValidator.php]: You are absolutely right about the `int`. But such callbacks usually return `-1`, `0` or `1`. Therefore I don't understand the `99`. Please also make it possible to test the diff.

Your current code will yield wrong results:

```
$validator = (new InArrayValidator())
    ->setHaystack(['a', 'b', 'c']);

var_dump($validator->isValid(['a', 'd']), $validator->getMessages());
```

The comparison functions can be static methods.

Comment (sukhwinder33445) [reply src/InArrayValidator.php]: Are the diff array keys also of interest? Can't find a use case for it. I assume we only want the values.

Comment (lippserd) [reply src/InArrayValidator.php]: Only values.

Comment (sukhwinder33445) [reply src/InArrayValidator.php]: the array_udiff comparison function returns unexpected results, especially with strict check. Discussed with @nilmerg and we decided to add the old variant again.

File: src/InArrayValidator.php [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * Computes the difference of the given values with the help of the haystack
+     *
+     * @param array $values
+     *
+     * @return array Values that was not found in the haystack
+     */
+    public function diffValues(array $values = []): array
```
Comment (lippserd) [inline src/InArrayValidator.php:117]: I would call this method something like `findInvalid`. `diffValues` is far too general for what it actually does, and you need PHPDoc to explain that.

File: tests/InArrayValidatorTest.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace ipl\Tests\Validator;
+
+use ipl\I18n\NoopTranslator;
+use ipl\I18n\StaticTranslator;
+use ipl\Validator\InArrayValidator;
+
+class InArrayValidatorTest extends TestCase
+{
+    public function testWithValidValue()
```
Comment (lippserd) [inline tests/InArrayValidatorTest.php:11]: You test what with valid value 😆? I would just reduce the method names to `testValidValues`, `testInvalidValues` and so on.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: src/InArrayValidator.php [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+     * ```
+     *  (new InArrayValidator())
+     *      ->setHaystack(['a', 'b', 'c', '295'])
+     *      ->setStrict()
+     *      ->findInvalid(['a', 'o', 295]); // Returns ['o', 295]
+     * ```
+     *
+     * @param array $values
+     *
+     * @return array Values that was not found in the haystack
```
Comment (lippserd) [inline src/InArrayValidator.php:124]: Either `Values not found in the haystack` or `Values that were not found in the haystack`. I prefer the first one.

File: src/InArrayValidator.php [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+                count($notInArray)
+            ),
+            implode(', ', $notInArray)
+        ));
+
+        return false;
+    }
+
+    /**
+     * Return the values that are not present in the haystack
```
Comment (lippserd) [inline src/InArrayValidator.php:111]: `Get the values from the specified array that are not present in the haystack`

File: tests/InArrayValidatorTest.php [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        $validator->setStrict(false);
+        $this->assertFalse($validator->isStrict());
+
+        $this->assertSame(['test', 'foo', 'bar', 5, [], 0.008], $validator->getHaystack());
+
+        $validator->setHaystack([]);
+        $this->assertSame([], $validator->getHaystack());
+    }
+
+    public function testMethodFindInvalid()
```
Comment (lippserd) [inline tests/InArrayValidatorTest.php:93]: `testFindInvalid`

**Review by lippserd — CHANGES_REQUESTED** [review]

**Review by lippserd — DISMISSED** [review]

File: src/InArrayValidator.php [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+     *      ->setHaystack(['a', 'b', 'c', '295'])
+     *      ->setStrict()
+     *      ->findInvalid(['a', 'o', 295]); // Returns ['o', 295]
+     * ```
+     *
+     * @param array $values
+     *
+     * @return array Values not found in the haystack
+     */
+    public function findInvalid(array $values = []): array
```
[the remark this answers is not in the document]
Comment (lippserd) [reply src/InArrayValidator.php:126]: We could do the following:

* Make it protected
* Extend it in the tests and change visibility to public
* Use the extended class in the tests

Not beautiful but I do the same in the simplify password element PR.

[the remark this answers is not in the document]
Comment (nilmerg) [reply src/InArrayValidator.php:126]: Yes, I like that.

[the remark this answers is not in the document]
Comment (nilmerg) [reply src/InArrayValidator.php:126]: *This is needed to request changes*

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]

**Review by lippserd — APPROVED** [review]

**Review by nilmerg — APPROVED** [review]

**Pull request merged**